### PR TITLE
feat: Added docstring examples to AnimatedBuilder and ChangeNotifier

### PR DESCRIPTION
**Pull request by albertodev01**
I have added an example to document how an `AnimatedBuilder` can be used along with a `ChangeNotifier`. Follow up for https://github.com/flutter/flutter/pull/97256

*List which issues are fixed by this PR. You must list at least one issue.*

 - https://github.com/flutter/flutter/issues/98624

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by albertodev01** [comment]
@Hixie I have only updated the documentation so no tests are required

File: packages/flutter/lib/src/widgets/transitions.dart [hunk 1]
```diff
@@ -993,6 +993,102 @@ class DefaultTextStyleTransition extends AnimatedWidget {
...
+/// To avoid unnecessary work, the [AnimatedBuilder] class can be used along
+/// with a [ChangeNotifier] to only rebuild certain portions a widget,
+/// leaving other parts and the subtree untouched.
+///
+/// Since a [ChangeNotifier] is a subtype of [Listenable], it can be used in
+/// an [AnimatedBuilder] to listen for changes. For example, to keep track
+/// of the state of a simple counter, a [ChangeNotifier] can be created and
+/// exposed to the widgets below using an [InheritedWidget].
+///
+/// {@tool snippet}
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/transitions.dart:1012]: These should probably use the `sample` or `dartpad` notation. That will subject them to analysis and testing, and the later will make them interactive examples in the docs.

Comment (albertodev01) [reply packages/flutter/lib/src/widgets/transitions.dart:1012]: I've just pushed a change where I removed `{@tool snippet}` because I saw it caused problems to the linux CI step

File: packages/flutter/lib/src/widgets/transitions.dart [hunk 2]
```diff
@@ -993,6 +993,99 @@ class DefaultTextStyleTransition extends AnimatedWidget {
...
+///
+/// This is an example that shows how an [InheritedWidget] can expose a
+/// [ChangeNotifier] to a widget and its children.
+///
+/// ```dart
+/// class Counter extends ChangeNotifier {
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/transitions.dart:1018]: This is re-implementing ValueNotifier. Can we just use that class to not re-invent the wheel?

File: packages/flutter/lib/src/widgets/transitions.dart [hunk 3]
```diff
@@ -993,6 +993,99 @@ class DefaultTextStyleTransition extends AnimatedWidget {
...
+///
+/// Since a [ChangeNotifier] is a subtype of [Listenable], it can be used in
+/// an [AnimatedBuilder] to listen for changes. For example, to keep track
+/// of the state of a simple counter, a [ChangeNotifier] can be created and
+/// exposed to the widgets below using an [InheritedWidget].
+///
+/// {@tool snippet}
+///
+/// This is an example that shows how an [InheritedWidget] can expose a
+/// [ChangeNotifier] to a widget and its children.
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/transitions.dart:1015]: Why show an example of that in the documentation of AnimatedBuilder?

Comment (goderbauer) [reply packages/flutter/lib/src/widgets/transitions.dart:1015]: Let's focus on what we actually want to show: How to use AnimatedBuilder with a ChangeNotifier.

File: packages/flutter/lib/src/widgets/transitions.dart [hunk 4]
```diff
@@ -993,6 +993,99 @@ class DefaultTextStyleTransition extends AnimatedWidget {
...
 ///
+/// {@template flutter.flutter.animatedbuilder_changenotifier.rebuild}
+/// ## Granular rebuilds with AnimatedBuilder and ChangeNotifier
+///
+/// A [State.setState] method call requires Flutter to entirely rebuild the
+/// current widget plus a variable portion of the subtree, depending on the
+/// depth and the structure.
+///
+/// To avoid unnecessary work, the [AnimatedBuilder] class can be used along
+/// with a [ChangeNotifier] to only rebuild certain portions a widget,
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/transitions.dart:1004]: Any Listenable works. Maybe list the ChangeNotifier as a concrete example, but we should also say that any Listenable will work.

File: packages/flutter/lib/src/widgets/transitions.dart [hunk 5]
```diff
@@ -993,6 +993,99 @@ class DefaultTextStyleTransition extends AnimatedWidget {
...
+///
+///   @override
+///   bool updateShouldNotify(CounterWidget oldWidget) {
+///     return counter.value != oldWidget.counter.value;
+///   }
+/// }
+/// ```
+/// {@end-tool}
+///
+/// In the following example, `CounterPage` is a child of `CounterWidget` so
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/transitions.dart:1055]: The examples should probably just be merged with a focus on the AnimationBuilder.

File: packages/flutter/lib/src/widgets/transitions.dart [hunk 6]
```diff
@@ -993,6 +993,99 @@ class DefaultTextStyleTransition extends AnimatedWidget {
...
+///   const CounterWidget({
+///     Key? key,
+///     required this.counter,
+///     required Widget child,
+///   }) : super(key: key, child: child);
+///
+///   final Counter counter;
+///
+///   static CounterWidget of(BuildContext context) {
+///     final result = context.dependOnInheritedWidgetOfExactType<CounterWidget>();
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/transitions.dart:1042]: please type all variables.

File: packages/flutter/lib/src/widgets/transitions.dart [hunk 7]
```diff
@@ -993,6 +993,99 @@ class DefaultTextStyleTransition extends AnimatedWidget {
...
+/// {@end-tool}
+///
+/// In the following example, `CounterPage` is a child of `CounterWidget` so
+/// it can use the context to retrieve the `Counter` instance. Whenever the
+/// button is pressed, only the [Text] inside the [AnimatedBuilder] is
+/// rebuilt while the `CounterPage` widget itself and its children aren't.
+///
+/// ```dart
+/// class CounterPage extends StatelessWidget {
+///   const CounterPage({Key? key,}) : super(key: key);
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/transitions.dart:1062]: remove comma.

File: packages/flutter/lib/src/widgets/transitions.dart [hunk 8]
```diff
@@ -993,6 +993,99 @@ class DefaultTextStyleTransition extends AnimatedWidget {
...
+///       floatingActionButton: FloatingActionButton(
+///         onPressed: () => CounterWidget.of(context).counter.value++,
+///         tooltip: 'Increment',
+///         child: const Icon(Icons.add),
+///       ),
+////    );
+///   }
+/// }
+/// ```
+/// {@endtemplate}
```
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/transitions.dart:1087]: We have started to extract sample code and put them into their own separate files so they are testable. Can you please do this with this example as well and add some tests for it? You can find an example of how to do that in https://github.com/flutter/flutter/pull/98440/files

**Comment by albertodev01** [comment]
Thank you @goderbauer for your time, I've created and tested an example as you suggested

**Comment by goderbauer** [comment]
Looks like the checks are unhappy with this change. Can you take a look?

File: examples/api/lib/foundation/change_notifier/change_notifier.0.dart [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+import 'package:flutter/material.dart';
+
+void main() {
+  runApp(const CounterApp());
+}
+
+/// This class holds the counter value and the text color.
+class Counter extends ChangeNotifier {
+  int _counter;
+  Color _textColor;
```
Comment (goderbauer) [inline examples/api/lib/foundation/change_notifier/change_notifier.0.dart:16]: style nit: place this between the getter and setter.

File: examples/api/lib/foundation/change_notifier/change_notifier.0.dart [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+
+  /// Creates a [Counter] class and, by default, sets [value] to `0` and
+  /// [textColor] to [Colors.black].
+  Counter({
+    int value = 0,
+    Color textColor = Colors.black,
+  }) : _counter = value, _textColor = textColor;
+
+  /// The counter value.
```
Comment (goderbauer) [inline examples/api/lib/foundation/change_notifier/change_notifier.0.dart:25]: most of the doc comments in this file don't provide any value because the don't cary any additional information over the name of the property. Maybe just remove them?

File: examples/api/lib/foundation/change_notifier/change_notifier.0.dart [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+  set textColor(Color colorValue) {
+    if (_textColor != colorValue) {
+      _textColor = colorValue;
+      notifyListeners();
+    }
+  }
+}
+
+/// An [InheritedWidget] that exposes a [Counter] class to the subtree.
+class CounterWidget extends InheritedWidget {
```
Comment (goderbauer) [inline examples/api/lib/foundation/change_notifier/change_notifier.0.dart:45]: maybe: CounterProvider instead of CounterWidget since it doesn't display a Counter.

**Comment by goderbauer** [comment]
Overall I am actually not convinced that this needs a huge example. Maybe a simple sentence in the doc would be enough. Something like: "Oh, by the way, an AnimiatedBuilder accepts any Listenable (not just an Animation) and works great with Listenable subclasses such as ChangeNotifier, ValueNotifier, etc." (of course phrased a little more formal)

That may be easier to understand than having to parse a big block of sample code.

File: examples/api/lib/foundation/change_notifier/change_notifier.0.dart [hunk 12]
```diff
@@ -0,0 +1,154 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
[the remark this answers is not in the document]
Comment (albertodev01) [reply examples/api/lib/foundation/change_notifier/change_notifier.0.dart:1]: Good idea, gonna do it!

**Comment by albertodev01** [comment]
@goderbauer I have simplified the text and also simplified the example.

I think that having a short example (along with some descriptive text) is a good thing. A person can play with the code, see it in action and we can make sure that our words are misunderstood 🙂 

I've followed you valuable comments to improve the code. I've also removed the `InheritedWidget` which yeah adds more complexity. It's all inside a simple `StatefulWidget` now and I agree that it looks simpler/better

**Comment by goderbauer** [comment]
Looks like one of the checks is failing. Can you take a look?

**Comment by albertodev01** [comment]
I've addressed all of your feedbacks and figure out the proper fix for the docs.

Thank you for your reviews

**Comment by goderbauer** [comment]
I have thought about this some more and I still think what I said in https://github.com/flutter/flutter/pull/98628#issuecomment-1045254459 is valid. A simple doc comment saying that any subclass of a Listenable (e.g. ChangeNotifier and friends) can be used with AnimatedBuilder is clearer and more concise than this example. You have to do a lot more reading to get that point from this example - and may even miss what the example is trying to illustrate altogether.

**Comment by albertodev01** [comment]
Would you suggest to simply drop the example and keep these lines?

```dart
/// A [State.setState] method call requires Flutter to entirely rebuild the
/// current widget plus a variable portion of the subtree, depending on the
/// depth and the structure.
///
/// To avoid unnecessary work, the [AnimatedBuilder] class can be used along
/// with a [Listenable] to only rebuild certain portions a widget,
/// leaving other parts and the subtree untouched. For example,
/// [ChangeNotifier] and [ValueNotifier] are subtypes of [Listenable].
///
/// Since a [ChangeNotifier] is a subtype of [Listenable], it can be used in
/// an [AnimatedBuilder] to listen for changes.
```

---

I would really like to keep the example but we can remove it if you want 🙂

I think that we could keep the lines I posted above. If we wrote even less contents, maybe it wouldn't give enough context to understand why one should use a ChangeNotifier in an AnimatedBuilder. Also, a very few lines wouldn't add much value and so they wouldn't even need to be there.

What do you think? Do you agree in _at least_ keeping that text?

**Comment by goderbauer** [comment]
Here's my suggestion for the docs:

```
## Improve rebuild performance using AnimationBuilder

Despite the name, [AnimationBuilder] is not limited to [Animation]s: Any subtype
of [Listenable] (e.g. [ChangeNotifier] and [ValueNotifier]) can be used with an
[AnimatedBuilder] to rebuild only parts of a widget when the [Listenable] notifies its listeners.
```

If you want to include an example below that, let's use one that is concise and focused on that particular use case. Here's my suggestion:

```
The following example implements a simple counter widget, that uses the
[AnimationBuilder] to limit rebuilds to only the [Text] widget showing the current 
count (stored in a [ValueNotifier]) when the value of the counter is changing.
```

```dart
import 'package:flutter/material.dart';

class CounterBody extends StatelessWidget {
  const CounterBody({Key? key, required this.counterValueNotifier}) : super(key: key);

  final ValueNotifier<int> counterValueNotifier;

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          const Text('Current counter value:'),
          // Thanks to the [AnimatedBuilder], only the widget displaying the
          // current count is rebuilt when `counterValueNotifier` notifies its
          // listeners.
          AnimatedBuilder(
            // [AnimatedBuilder] accepts any [Listenable] subtype.
            animation: counterValueNotifier,
            builder: (BuildContext context, Widget? child) {
              return Text('${counterValueNotifier.value}');
            },
          ),
        ],
      ),
    );
  }
}

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final ValueNotifier<int> _counter = ValueNotifier<int>(0);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(title: const Text('AnimatedBuilder example')),
        body: CounterBody(counterValueNotifier: _counter),
        floatingActionButton: FloatingActionButton(
          onPressed: () => _counter.value++,
          child: const Icon(Icons.add),
        ),
      ),
    );
  }
}

void main() {
  runApp(const MyApp());
}
```

**Comment by albertodev01** [comment]
I really like your changes! Thank you for having considered keeping an example, I think it will be very valuable

File: dev/bots/allowlist.dart [hunk 13]
```diff
@@ -25,7 +25,6 @@ const Set<String> kCorePackageAllowList = <String>{
   'boolean_selector',
   'characters',
   'charcode',
-  'cli_util',
```
Comment (jmagman) [inline dev/bots/allowlist.dart:28]: This should be done in its own PR https://github.com/flutter/flutter/pull/99037 as it's totally unrelated to your docstring change.
Please hold off merging this until https://github.com/flutter/flutter/pull/99037 lands (it should merge cleanly so you don't need to update this PR).

Comment (goderbauer) [reply dev/bots/allowlist.dart:28]: I'd still suggest rebasing this PR to the current master to make this diff go away. It will look odd for future archeologists looking at this PR.

**Review by goderbauer — APPROVED** [review]
LGTM

Please rebase it to the latest master, though.

**Comment by albertodev01** [comment]
I have rebased but the diff isn't going away...

File: dev/bots/allowlist.dart (hunk 13 above)
Comment (goderbauer) [reply dev/bots/allowlist.dart:28]: https://github.com/flutter/flutter/pull/99037 has been submitted, so this PR should be good to go.

**Review by goderbauer — APPROVED** [review]
LGTM

**Review by Piinks — APPROVED** [review]
Flutter_LGTM
